### PR TITLE
perf: bucket smoothness regularization

### PR DESCRIPTION
**Pull request by jorgenhanssen**
## Summary

Added a bucket smoothness regularization term to the NNUE training loss. The motivation came from watching the eval during games. It would sit around +1.2, jump to +2.0, stay there for a while, then drop back to +1.2. Very steppy. I figured the output heads for adjacent game-phase buckets were learning pretty different functions, and the eval was just snapping between them at the transition points.

This is probably bad for search quality too, since the `improving` heuristic used for pruning decisions relies on the eval changing smoothly. if there are big artificial cliffs between buckets, it's going to misfire.

I tried adding a simple loss to penalize squared differences between adjacent bucket outputs. This nudges neighboring buckets to agree with each other without forcing them to be identical, so the eval transitions more gently across game phases.

And it worked surprisingly well actually. I trained some simple configs with different smoothness loss for 3 epochs on the 500m dataset used to train `1.1.0`, and:

```
--------------------------------------------------
Rank Name                             Elo        +/-       nElo        +/-      Games      Score       Draw           Ptnml(0-2)
   1 grail0.1                       30.77      18.94      45.45      27.80        600      54.4%      40.7% [12, 56, 122, 87, 23]
   2 grail0.2                        2.32      19.55       3.30      27.80        600      50.3%      36.3% [15, 81, 109, 75, 20]
   3 grail0.3                       -6.95      18.93     -10.22      27.80        600      49.0%      41.3% [20, 71, 124, 71, 14]
   4 grail0.0                      -26.11      18.92     -38.54      27.80        600      46.2%      38.3% [21, 89, 115, 64, 11]
--------------------------------------------------
```

So that is pretty neat. I will re-train with `0.1` until early stop using the same data and check against `next`

### CODE REVIEW

File: nnue/src/network/model.rs [hunk 1]
```diff
@@ -6,55 +6,58 @@ use crate::encoding::NUM_FEATURES;
...
-        let embedding_out = x.apply(&self.embedding)?.relu()?;
+    /// Forward pass returning all bucket outputs: [batch, OUTPUT_BUCKETS].
+    pub fn forward(&self, x: &Tensor) -> Result<Tensor> {
+        let embedding = x.apply(&self.embedding)?.relu()?;
+        self.eval_head.forward(&embedding)
+    }
+}
+
+/// Phase-specific evaluation head with one output stack per bucket.
+pub struct EvalHead {
```
Comment (jorgenhanssen) [inline nnue/src/network/model.rs:30]: unrelated refactor. i want to try some sort of policy head at some point, so this is prepping for that. i also like it architecturally, even if it is a bit over-engineered atm.

**Comment by jorgenhanssen** [comment]
Nah, fun POC but not holding up after a full training:

```
--------------------------------------------------
Results of grail vs grail-next (10+0.1, 1t, 256MB, UHO_Lichess_4852_v1.epd):
Elo: -81.19 +/- 38.89, nElo: -134.08 +/- 61.65
LOS: 0.00 %, DrawRatio: 47.54 %, PairsRatio: 0.23
Games: 122, Wins: 23, Losses: 51, Draws: 48, Points: 47.0 (38.52 %)
Ptnml(0-2): [8, 18, 29, 6, 0], WL/DD Ratio: 1.42
--------------------------------------------------
```
